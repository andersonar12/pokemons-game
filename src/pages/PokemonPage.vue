<template>
  <div class="pokemon-page">
    <h1>Quien es este pokemon??</h1>
    <div v-if="pokemon">
      <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showImage" />
      <PokemonOptions :pokemons="pokemonsArray" @on-selection="(e) => checkAnswer(e)" />

      <div v-if="message">
        <h2>{{ message }}</h2>
        <button @click="tryAgain">Siguiente intento</button>
      </div>
    </div>

    <h1 v-else>Espere a que cargue el pokemon...</h1>
  </div>
</template>

<script>
import PokemonPicture from "../components/PokemonPicture.vue";
import PokemonOptions from "../components/PokemonOptions.vue";
import getPokemonOptions from "../functions/getPokemonOptions";

export default {
  name: "PokemonPage",
  components: {
    PokemonPicture,
    PokemonOptions,
  },
  data() {
    return {
      pokemonsArray: [],
      pokemon: null,
      showImage: false,
      message: "",
    };
  },
  mounted() {
    this.mixPokemonArray();
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonsArray = await getPokemonOptions();
      const randomInt = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemonsArray[randomInt];
    },
    checkAnswer(answer) {
      this.showImage = true;
      const name = this.pokemon.name.charAt(0).toUpperCase() + this.pokemon.name.slice(1);
      if (answer.id === this.pokemon.id) {
        this.message = `Correcto, ${name}`;
      } else {
        this.message = `Oops, era ${name}`;
      }
    },
    tryAgain() {
      this.showImage = false;
      this.pokemon = null;
      this.mixPokemonArray();
      this.message = "";
    },
  },
  computed: {},
  watch: {},
};
</script>

<style scoped>
.pokemon-page {
}
</style>
