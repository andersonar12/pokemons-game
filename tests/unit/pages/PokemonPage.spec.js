import { shallowMount, mount } from "@vue/test-utils";
import PokemonPage from "@/pages/PokemonPage";
import { mockPokemons } from "../mocks/pokemons.mocks.js";

describe("PokemonPage Component", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage);
  });

  test("debe de hacer match con el snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("debe de llamar mixPokemonArray al montar", () => {
    const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, "mixPokemonArray");
    const wrapper = shallowMount(PokemonPage);

    expect(mixPokemonArraySpy).toHaveBeenCalled();
  });

  //   test("debe de hacer match con el snapshot cuando cargan los pokemons", () => {
  //     const wrapper = shallowMount(PokemonPage, {
  //       data() {
  //         return {
  //           pokemonsArr: mockPokemons,
  //           pokemon: mockPokemons[0],
  //           showPokemon: false,
  //           showAnswer: false,
  //           message: "",
  //         };
  //       },
  //     });

  //     expect(wrapper.html()).toMatchSnapshot();
  //   });

  //   test("debe de mostrar los componentes de PokemonPicture y PokemonOptions", () => {
  //     const wrapper = shallowMount(PokemonPage, {
  //       data() {
  //         return {
  //           pokemonsArray: mockPokemons,
  //           pokemon: mockPokemons[0],
  //           showPokemon: false,
  //           showAnswer: false,
  //           message: "",
  //         };
  //       },
  //     });

  //     // console.log(wrapper.html());
  //     const picture = wrapper.find("pokemon-picture-stub");
  //     const options = wrapper.find("pokemon-options-stub");

  //     expect(picture.exists()).toBeTruthy();
  //     expect(options.exists()).toBeTruthy();

  //     expect(picture.attributes("pokemonid")).toBe("1");
  //     expect(options.attributes("pokemons")).toBeTruthy();
  //   });

  test("pruebas con checkAnswer", async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonsArray: mockPokemons,
          pokemon: mockPokemons[0],
          showImage: false,
          message: "",
        };
      },
    });

    await wrapper.vm.checkAnswer(0);

    // console.log(wrapper.html());

    expect(wrapper.find("h2").exists()).toBeTruthy();
    expect(wrapper.vm.showImage).toBe(true);
    expect(wrapper.find("h2").text()).toBe(`Correcto, ${mockPokemons[0].name}`);

    await wrapper.vm.checkAnswer(10);

    expect(wrapper.vm.message).toBe(`Oops, era ${mockPokemons[0].name}`);
  });
});
