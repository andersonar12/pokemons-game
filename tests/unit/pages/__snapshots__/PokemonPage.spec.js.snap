// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component debe de hacer match con el snapshot 1`] = `
<div class="pokemon-page">
  <h1>Quien es este pokemon??</h1>
  <h1>Espere a que cargue el pokemon...</h1>
</div>
`;

exports[`PokemonPage Component debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div class="pokemon-page">
  <h1>Quien es este pokemon??</h1>
  <div>
    <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
    <pokemon-options-stub pokemons=""></pokemon-options-stub>
    <!--v-if-->
  </div>
</div>
`;
